### ISSUE TITLE
Alt text validation pattern refinement

### ISSUE THREAD

**Issue by thibaudcolas** [wagtail/wagtail#12091]
### Is your proposal related to a problem?

In #11967, we introduced Wagtail’s first ever custom accessibility check, `alt-text-quality`. It checks all images’ alt text to flag known bad patterns. Currently it uses this regex: `\.(avif|gif|jpg|jpeg|png|svg|webp)$`.

There’s room for improvement here, there are a lot of other known bad patterns which we could flag.

### Describe the solution you'd like

Extend the regular expression to flag more known bad patterns in alt text. Focusing on patterns that are very clear-cut problems, in particular patterns that would fail [WCAG SC 1.1.1 Non-text Content](https://www.w3.org/TR/2008/REC-WCAG20-20081211/#text-equiv-all). For example:

- Alt text starting with `IMG_` is probably a reasonable indication that most of the alt text is a file name, and as such doesn’t serve an "equivalent purpose" to the image’s visuals.
- Alt text starting with "Image of" is more verbose than it needs to be, but otherwise is an indication that the alt text is likely a description of the image, which is _good_

### Describe alternatives you've considered

See [RFC 89: ATAG 2.0 audit](https://github.com/wagtail/rfcs/pull/89), [RFC 51: Contextual alt text](https://github.com/wagtail/rfcs/pull/51), [RFC 97: Alt Text Capabilities](https://github.com/wagtail/rfcs/pull/97).

### Additional context

I have prepared two datasets of images alt text which we can use to trial our new rule:

- [35k alt text of images](https://docs.google.com/spreadsheets/d/1MgHVWYuKjdLkqiCOJzli7wQKjmjpeHjLL1naJYHeeDw/edit#gid=2009276316) –  displayed on Wagtail websites’ homepages
- [19k alt text of CMS-managed images](https://docs.google.com/spreadsheets/d/1MgHVWYuKjdLkqiCOJzli7wQKjmjpeHjLL1naJYHeeDw/edit#gid=303545789) – also from Wagtail websites’ homepages

See [thibaudcolas/alt-text-benchmark](https://github.com/thibaudcolas/alt-text-benchmark) for a reference on how this data was produced.

### Working on this

Contributions welcome from anyone with thoughts on what best to aim for in the context of Wagtail. @albinazs will implement the one code change and additional unit tests once we’ve decided on a new pattern.

**Comment by streats** [comment]
> an indication that the alt text is likely a description of the image, which is good

Just checking I understood - are we saying that this would be a good alt text? My understanding is that "Image of" and "Photo of" are unnecessary as AT announces it's an image, but "Screenshot of" may be appropriate.

**Comment by streats** [comment]
Some other ideas off the top of my head, with examples from the 35k file:

- Multiple words with no spaces but other common separators like `-` or `_`. These are unlikely to be proper sentences. 
   - Examples: `acm-phase-4`, `social-icon-instagram`
- Single words. While it might be possible that alt text could have a single word and still be sufficient (for example, "Beyonce", "Kitten" or "Facebook"), it's unlikely to be good quality. 
   - Examples: `play`, `logo`, `media`
-  Starts with lowercase. Again there might be legitimate use cases ("bell hooks in October 2014" or "iPhone 15 launch event") but generally this could indicate poor quality. 
   - Examples: `leadership programs square graphic`, `gratitude daily tile`, `wsj`
- All uppercase. While there may be legitimate inclusion of acronyms or initialisms, an entire alt text of uppercase is unlikely to have been written by a human, or if it was it's probably minimally descriptive. 
  - Example: `AOC`, `DSC00752`, `FICEP`
- Contains only numbers. I can't think of a legitimate reason why alt text would contain only numbers unless it was an image of text, which would be inaccessible for its own reasons. 
   - Examples: `110033`, `2017`
- Starts with a non-alphanumeric character. There may be legitimate exceptions such as Spanish `¡ {sentence} !` (though I don't think exclamation marks are necessarily appropriate in alt text)
  - Examples: ` + alt_text + '`, `_CAM5839`, `(( title ))`
- Emoji. 
  - Example: `🍃 Chicken Wedge Salad`, `✨ Carved mirrors and beautiful paintings converge`

**Comment by thibaudcolas** [comment]
@streats at this stage we only have the capability to flag things either as errors or not at all, there is no third "review" option. So we want to flag "definitely bad" alt text, not "possibly could be improved". So any proposed pattern needs to be as close to 99% precision as possible (high quality of matches). Having a high recall (high quantity) isn’t needed because we don’t have the UI to display the nuance of possible issues.

Also keep in mind we want patterns that work across multilingual content, so rules that flag non-alphanumeric characters aren’t viable.

---

> Just checking I understood - are we saying that this would be a good alt text? My understanding is that "Image of" […]

Yes, your understanding is correct. What I’m saying is if alt text starts with "Image of", then we can reasonably expect what follows will be a description of the image. The alt text would clearly be better without the "Image of" prefix, but that’s a minor verbosity issue. The prefix won’t prevent a screen reader user from understanding the content. And I don’t think the presence of such a verbose prefix would constitute a fail of [WCAG SC 1.1.1 Non-text content](https://www.w3.org/TR/2008/REC-WCAG20-20081211/#text-equiv-all).

> Multiple words with no spaces but other common separators like `-` or `_`

For underscores this makes sense to me. For `-`, how would you distinguish between problem cases and legitimate hyphenation? For example "Coca-Cola", or the French word for Rainbow ("Arc-en-Ciel") would make for an excellent alt text. Or place names like Stoke-on-Thames.

Also I believe screen readers have no problem reading hyphenated text without mention of the hyphens, since there are a lot of words like this. For example "Coca-Cola" is read "Coca Cola", not "Coca Hyphen Cola".

> Single words (contains no spaces).

I don’t see how this would work as a global rule with the diversity of images we work with in a CMS? "Facebook" seems like perfectly valid alt text for an image of the text "Facebook".

Also keep in mind we need patterns that work across different languages. In some languages there are much longer words that could be fine to use as-is even for longer description. For example in Finnish, "Pyyhkäisyelektronimikroskooppi".

> Starts with lowercase.

I think just "iPhone" is a solid example of why this couldn’t work. There’s lots of text out there that’s stylized as lowercase.

> All uppercase. While there may be legitimate inclusion of acronyms or initialisms, an entire alt text of uppercase is unlikely to have been written by a human, or if it was it's probably minimally descriptive.

This seems to me like a very common occurrence for logos ("NHS", "FCC", "MIT")

> Contains only numbers.

I think this is promising but there are clear examples where numbers-only could be valid so I’m hesitant. WCAG has clear provisions where [images of text](https://www.w3.org/WAI/WCAG22/Understanding/images-of-text.html) are allowed: _A particular presentation of text is essential to the information being conveyed._, and mentions logotypes are valid examples.

So here are examples of things that are referred to with numbers only: [2012 (film)](https://en.wikipedia.org/wiki/2012_(film)), [1664 (beer)](https://en.wikipedia.org/wiki/1664_(beer)), [1917 (2019 film)](https://en.wikipedia.org/wiki/1917_(2019_film)).

> Starts with a non-alphanumeric character.

This would need refinement for alphabets that don’t use alphanumeric characters. Perhaps if we earmarked specific characters?

> Emoji. Even if the alt text is good, […]

I can see why alt text would be better without emoji, but it seems to me like a clear case of "could use improvement" rather than "alt text is bad".

> Multiple consecutive non-alphanumeric characters.

`What&#x27;s The Point Of Literature?` is just HTML escaping of a `'` single quote character as the template rendering logic might not know whether the developer chose to use single or double quotes for their attribute. Browsers have no problem interpreting this as "'" when providing the alt text to a screen reader.

**Comment by thibaudcolas** [comment]
> Perhaps those characters aren’t alphanumeric from a technical standpoint?

Gotcha, yes, I suppose it depends what we mean by alphanumeric :) I thought you meant ASCII characters, but yes alphanumeric in Unicode would include accented letters and I assume much more than that. This gets quite thorny so if we did something like this I assume we’d need to review what is feasible in JavaScript specifically (which is where our regex check runs currently).

> an allow list for legitimate characters or a denylist for illegitimate ones?

Yes so same for this I think it depends on what is technically feasible. So we’d need to look at available [character class escapes](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Regular_expressions#escape_sequences) specifically, narrowing down on the ones that are unicode-aware. Possibly a [unicode character class escape](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Regular_expressions/Unicode_character_class_escape) with the correct character classes 🤔.

**Comment by thibaudcolas** [comment]
Noting from today’s accessibility team meeting – for the pass/fail checks, we’re keen to proceed with:

- `_` – "any underscore in the alt text = error". We noted there would be a few false positives (brands with underscores, images of computer code, images of social media handles or URLs with underscores. Those feel like they’re reasonably uncommon enough, and problematic anyway as images of text, to justify the trade-off.
- TBC - single-character alt text (`^.$`). There are a few characters that seem like they could be appropriate (particularly for icons – `🙂`, `”`, `→`). We want to do more research and then hopefully proceed.

**Comment by thibaudcolas** [comment]
I’ve merged #12119 with the inclusion of the underscore. Keeping this open for us to do further research into other patterns, such as single-character alt text. It seems like we could disallow specific unicode or character ranges (for example `^[0-9a-z]$` would probably go a long way).

**Comment by thibaudcolas** [comment]
Re-reviewing this, I expect we’ve done enough on this for now. There is room to refine further for single-character alt text, but that’s just one of many problem patterns. The possible returns are diminishing fast. It’s now better documented how to customize those check patterns, for projects wanting fuller checking.
